relentless/core/variable.py

```diff
@@ -0,0 +1,395 @@
+        try:
+            if high is None or self._high is None:
+                self._high = high
+            elif high > self._low:
+                self._high = high
+            else:
+                raise ValueError('The high bound must be greater than the low bound')
+        except AttributeError:
+            self._high = high
+
+        try:
+            self._value, self._state = self.clamp(self._value)
+        except AttributeError:
+            pass
+
+    @property
+    def state(self):
+        """:py:class:`DesignVariable.State`: The state of the variable"""
+        return self._state
+
+    def isfree(self):
+        """Confirms if the variable is unconstrained or within the bounds.
+
+        Returns
+        -------
+        bool
+            True if the variable is unconstrained or within the bounds, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.FREE
+
+    def atlow(self):
+        """Confirms if the variable is at the lower bound.
+
+        Returns
+        -------
+        bool
+            True if the variable is at the lower bound, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.LOW
+
+    def athigh(self):
+        """Confirms if the variable is at the upper bound.
+
+        Returns
+        -------
+        bool
+            True if the variable is at the upper bound, False otherwise.
+
+        """
+        return self.state is DesignVariable.State.HIGH
+
+class DependentVariable(Variable):
+    """Abstract base class for a variable that depends on other variables.
+
+    A dependent variable is composed from other variables. In addition to the
+    value property, it must also supply a derivative method with respect to
+    any of its dependent variables.
+
+    The number of dependencies of the variable are locked at construction, but
+    their values can be modified.
+
+    Parameters
+    ----------
+    vardicts : dict
+        Attributes for the variable on which the DependentVariable depends (positional argument).
+    kwvars : kwargs
+        Attributes for the variable on which the DependentVariable depends (keyword arguments).
+
+    """
+    def __init__(self, *vardicts, **kwvars):
+        attrs = {}
+        for d in vardicts:
+            attrs.update(d)
+        attrs.update(**kwvars)
+
+        if len(attrs) == 0:
+            raise AttributeError('No attributes specified for DependentVariable.')
+
+        for k,v in attrs.items():
+            super().__setattr__(k,self._assert_variable(v))
+        self._params = tuple(attrs.keys())
+
+    def __setattr__(self, name, value):
+        #Sets the value of the variable on which the specified DependentVariable depends.
+        if name != '_params' and name in self._params:
+            value = self._assert_variable(value)
+        super().__setattr__(name,value)
+
+    @property
+    def depends(self):
+        """Generator of parameter,object pairs."""
```

Comment:
Can you expand this documentation a bit to explain what is returned? The first element is the string name of the parameter, and the second element is the object reference itself.